File: Eunleelee/2024-02/2월 23일/2개 이하로 다른 비트.py
def solution(numbers):
    answer = []

    def toBinary(number):
        ans = ''
        while number > 0:
            ans += str(number%2)
            number = number//2

        add_zero = 4-(len(ans)%4)
        ans += '0' * add_zero

        return ans[::-1]

    def toDecimal(number):
        ans = 0; l = len(number)-1
        for n in number:
            ans += (2**l)*int(n)
            l -= 1

        return ans

    for num in numbers:
        if num%2 == 0:
            answer.append(num+1)
        else:
            odd_to_binary = toBinary(num)

            for i in range(len(odd_to_binary)-1, -1, -1):
                if odd_to_binary[i] == '0':
                    odd_to_binary = odd_to_binary[:i] + '10' + odd_to_binary[i+2:]
                    answer.append(toDecimal(odd_to_binary))
                    break

    return answer

File: jja6312/2월23일/2개이하로다른비트.java
public class Solution {
    public long[] solution(long[] numbers) {
        long[] answer = new long[numbers.length];
        
        for(int i = 0; i < numbers.length; i++) {
            // 홀수인 경우
            if((numbers[i] & 1) == 1) {
                long lastZero = ~numbers[i] & (numbers[i] + 1);
                numbers[i] |= lastZero; // 가장 낮은 0 비트를 1로 변경
                numbers[i] &= ~(lastZero >> 1); // 변경된 비트 바로 다음의 1을 0으로 변경
            } else {
                // 짝수인 경우, 가장 낮은 자리의 비트를 1로 변경하기만 하면 됩니다.
                numbers[i] |= 1;
            }
            answer[i] = numbers[i];
        }
        
        return answer;
    }
    
    public static void main(String[] args) {
        Solution solution = new Solution();
        long[] numbers = {2, 7};
        long[] answers = solution.solution(numbers);
        
        for(long answer : answers) {
            System.out.print(answer + " ");
        }
    }
}
